### ISSUE TITLE
Refine -H:ReportAnalysisForbiddenType DevXp

### ISSUE THREAD

**Issue by sdeleuze** [oracle/graal#4853]
`-H:ReportAnalysisForbiddenType=Foo` is a super useful `native-image` option to find why a class `Foo` is included in a native image due to the reachability stack trace printed, Spring team is using it extensively.

That said, for some use cases, sometime no reachability stack trace is printed and that could be for various reasons:
 - I think for `java.*` classes and maybe some others, no stack trace is printed because those classes are not instrumented (to be confirmed via @gradinac for example)
 - If the class is included via reflection specified in `reflection-config.json`

The DevXp would be significantly improved if the reason why no stacktrace can be provided would be mentioned.

**Comment by fniephaus** [comment]
I believe this is following up on #4839. AFAIK instrumenting JDK classes does not always work at the moment, @olpaw would know for sure. IIRC the `-H:ReportAnalysisForbiddenType` option fails on first match, which is not a good DevXP to begin with. I wonder: wouldn't it be easier to use `-H:+PrintAnalysisCallTree` and then search for the types you consider forbidden in the output file? That would allow you to search for more than one forbidden type at a time.

We haven't figured out a good DevXP for these types of questions, so it'd be great if you could elaborate on your use cases a bit more.

**Comment by sdeleuze** [comment]
I think they are complementary. We extensively uses `-H:+PrintAnalysisCallTree` to search forbidden types, and it works fine. Where `-H:ReportAnalysisForbiddenType` really shines is to find the reachability stack trace which allows to understand how and why a type got included, hence the ask to provide an explanation when no stack trace is provided.

**Comment by olpaw** [comment]
> wouldn't it be easier to use -H:+PrintAnalysisCallTree and then search for the types you consider forbidden in the output file

IMO, `-H:+PrintAnalysisCallTree` is inferior to `-H:ReportAnalysisForbiddenType` because it will show a random path to the Type that you are interested in, but there can be more than one path that causes a Type to become reachable.

`-H:ReportAnalysisForbiddenType` on the other hand will show immediately when your type of interest gets reachable thus making it easier to understand what causes a type to get reachable and to hunt down all causes of reachability.

**Comment by fniephaus** [comment]
@olpaw makes sense. Now I wonder: we already have an infrastructure that allows marking something reachable when something else becomes reachable (for example used in conditional config). Could we possibly re-use that infrastructure and instead of a "mark something else as reachable" handler, we install a "collect reachability trace for this and add to a list" and then print that list at the end of the analysis if non-empty?

**Comment by fniephaus** [comment]
Had another chat with @olpaw about this. How about a new `-H:PrintReachabilityTraces` option that takes a list of classes or methods and outputs a reversed call tree with all the traces that make the classes or methods reachable?
